kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/MessageFetchContext.java

```diff
@@ -154,6 +155,37 @@ public void recycle() {
                             long offset = ((FetchRequest) fetchRequest.getRequest()).fetchData()
                                 .get(topicPartition).fetchOffset;
 
+                            // handle offset out-of-range exception
+                            ManagedLedgerImpl managedLedger = (ManagedLedgerImpl) tcm.getManagedLedger();
+                            try {
+                                long logStartOffset = MessageIdUtils.getLogStartOffset(managedLedger);
```

Comment:
`getLogStartOffset` will read entry from BookKeeper, which is time consuming, and it will called by each handle fetch request. It is too expensive.  In my opinion, after get exception, we try to call `getLogStartOffset` maybe better.

Reply:
I agree with @hangc0276 . I think we can only check `offset > logEndOffset` here.

And could you give an example to show in which case will the offset be less than `logStartOffset`? If it's hard to simulate this case, could you explain what happens if `offset < logStartOffset`? If it there will be an exception in `ManagedCursor#asyncReadEntries`, I think we can just add the check here.

Reply:
For example I have a flink job using KoP and I trigger a savepoint `S` at offset `A` and then stop it. After a period of time, the offset becomes to range `[A+N, A+M]` satisfying `N,M>0, M>N` due to topic retention. At this time the user chooses to recover its job from savepoint `S`, which will lead a out-of-range that meets `offset < logStartOffset`. Here is also a reference to the implementation of Kafka.

Reply:
Maybe we can utilize the short-circuit rules in logic operations to make some optimizations? 🤣 
```java
...
if (offset > logEndOffset || offset < MessageIdUtils.getLogStartOffset(managedLedger)) {...}
...
```

Reply:
if offset is between LSO and LEO range, it still go into `getLogStartOffset`

Reply:
So could you answer the question that what happens if `offset < logStartOffset`? If it's possible, could you add a test to cover this case?

Reply:
Agree with that `getLogStartOffset` is an expensive operation since it will goes into bookie.
An optimization for  reducing the cost of `getLogStartOffset`  is  adding a `startOffset` field  for metadata of ManagedLedger in broker side and then we can get the `startOffse` from ManagedLedger directly without reading entry from bookie.

Reply:
@aloyszhang It's okay. But it should be merged into pulsar repo ASAP because 2.8.0 is in releasing process now.

Reply:
@aloyszhang  We have another method to achieve this goal.